### ISSUE TITLE
MS1MV3 training rec file pre-shuffle causes strange cyclic loss pattern

### ISSUE THREAD

**Issue by zzzzz314314** [deepinsight/insightface#2059]
Hello,

I am training [Resnet34, MS1MV3, Arcface]. Since you have kindly provided the exact pretrained weights of the backbone, I only need to first freeze the pretrained backbone train the FC layer, and then finetune the whole network with the Arcface loss and my other custom losses.

During the training of the first stage (freeze pretrained backbone, train FC), I only use the [MXFaceDataset](https://github.com/deepinsight/insightface/blob/babb9a58bbc42ae4b648acdbb803159a35f53db3/recognition/arcface_torch/dataset.py#L39) but do not use dali, DistributedSampler, or DataLoaderX to wrap the  MXFaceDataset since I only have one machine. A simple torch Dataloader is used to wrap the MXFaceDataset.

During the training of the first stage (freeze pretrained backbone, train FC), I found the shuffling in the Dataloader causes far too much time (possibly because I only have hard disks and random accessing the rec file is time-consuming). Hence, I run the [rec2shufrec.py](https://github.com/deepinsight/insightface/blob/master/recognition/arcface_mxnet/common/rec2shufrec.py) to do a pre-shuffling so I can set the shuffle arg in Dataloader False.

To be compatible with the code in [MXFaceDataset](https://github.com/deepinsight/insightface/blob/babb9a58bbc42ae4b648acdbb803159a35f53db3/recognition/arcface_torch/dataset.py#L149), I do a bit change in the [rec2shufrec.py](https://github.com/deepinsight/insightface/blob/master/recognition/arcface_mxnet/common/rec2shufrec.py). To be specific, MXRecordIO is replaced with MXIndexedRecordIO to produce both idx and rec files. That's all of the change.

The training of the FC layer (93431 classes output) goes very fast and smooth. However, at the very middle of the second epoch, the loss surges. The same case happens at the middle of the third, forth, epochs, causing a strange cyclic loss pattern.

I have checked the learning rate, different optimizers, the images and labels after running the [rec2shufrec.py](https://github.com/deepinsight/insightface/blob/master/recognition/arcface_mxnet/common/rec2shufrec.py) shuffling. Nothing went wrong. Then I redo the [rec2shufrec.py](https://github.com/deepinsight/insightface/blob/master/recognition/arcface_mxnet/common/rec2shufrec.py) many times to create different training data streams, but the loss pattern persists. The cross entropy loss always surges at the very middle of each epochs. Hence I suspect something went wrong in the [rec2shufrec.py](https://github.com/deepinsight/insightface/blob/master/recognition/arcface_mxnet/common/rec2shufrec.py) and the use of [MXFaceDataset](https://github.com/deepinsight/insightface/blob/babb9a58bbc42ae4b648acdbb803159a35f53db3/recognition/arcface_torch/dataset.py#L39) though I cannot identify anything suspicious in the codes. Could you please give some clue on this? Many thanks.

(40k counts one epoch, at the first epoch, LR=1e-3, after that, LR=1e-5. No cyclic LR)
![image](https://user-images.githubusercontent.com/54563183/181166034-0dfd59b7-3c4c-4489-8b44-e4a4f7a00d98.png)

**Comment by jacqueline-weng** [comment]
I think if you only shuffle the dataset once before training, it is equivalent to not shuffling it at all. Data is fed into the model with a certain sequence for every epoch. Maybe you can use a random sample for each iteration.

**Comment by zzzzz314314** [comment]
Thanks for the reply. The data sequence of the original Ms1mv3 rec file is organized in the label order, and hence the preshuffle does make a difference. However, it is true that the preshuffling makes the data sequence identical in each epoch. For the mentioned random sampling, it may take lots of time due to random accessing the rec file and the absence of ssd. Am I understanding the point right? Thanks again.

**Comment by zzzzz314314** [comment]
Enabling the shuffling in the Dataloader solves the issue. I understand that the training data should be sampled from an iid distribution, and hence using the same random shuffled data sequence for each epoch might cause some (minor) problems. However, I am surprised that the effect can be so prominent despite that the dataset is so large and neural networks do not have sort of ability to memorize the optimization process. As such, personally I think one should reconsider if the training tips to save time as mentioned [here](https://github.com/deepinsight/insightface/issues/1426) is a valid idea.
Many thanks anyway.
